server/adaptors/integrations/repository/repository.ts

```diff
@@ -3,44 +3,66 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import * as path from 'path';
 import { IntegrationReader } from './integration_reader';
-import { FileSystemCatalogDataAdaptor } from './fs_data_adaptor';
 import { CatalogDataAdaptor } from './catalog_data_adaptor';
 
 export class TemplateManager {
-  reader: CatalogDataAdaptor;
-  directory: string;
+  readers: CatalogDataAdaptor[];
 
-  constructor(directory: string, reader?: CatalogDataAdaptor) {
-    this.directory = directory;
-    this.reader = reader ?? new FileSystemCatalogDataAdaptor(directory);
+  constructor(readers: CatalogDataAdaptor[]) {
+    this.readers = readers;
   }
 
   async getIntegrationList(): Promise<IntegrationReader[]> {
-    // TODO in the future, we want to support traversing nested directory structures.
-    const folders = await this.reader.findIntegrations();
+    const lists = await Promise.all(
+      this.readers.map((reader) => this.getReaderIntegrationList(reader))
+    );
+    const flattened = lists.flat();
+
+    // If there are collisions by name, prioritize earlier readers over later ones.
+    const seen = new Set();
+    return flattened.filter((item) => {
+      if (seen.has(item.name)) {
+        return false;
+      }
+      seen.add(item.name);
+      return true;
+    });
+  }
+
+  private async getReaderIntegrationList(reader: CatalogDataAdaptor): Promise<IntegrationReader[]> {
+    const folders = await reader.findIntegrations();
     if (!folders.ok) {
-      console.error(`Error reading integration directories in: ${this.directory}`, folders.error);
       return [];
     }
     const integrations = await Promise.all(
-      folders.value.map((i) =>
-        this.getIntegration(path.relative(this.directory, path.join(this.directory, i)))
-      )
+      folders.value.map((integrationName) => this.getReaderIntegration(reader, integrationName))
     );
     return integrations.filter((x) => x !== null) as IntegrationReader[];
   }
 
-  async getIntegration(integPath: string): Promise<IntegrationReader | null> {
-    if ((await this.reader.getDirectoryType(integPath)) !== 'integration') {
-      console.error(`Requested integration '${integPath}' does not exist`);
+  async getIntegration(integrationName: string): Promise<IntegrationReader | null> {
+    const maybeIntegrations = await Promise.all(
+      this.readers.map((reader) => this.getReaderIntegration(reader, integrationName))
```

Comment:
I'm not quite familiar with how the entire index adaptor works, but just wondering if there could have any performance issues as I saw there are recursive file traversals as well as file system operations. How did we perform performance tests and would this be a potential issue to the overall cluster health if the cluster is undergoing high traffic with less computing resource available.

Reply:
Short answer is that it's not very performant: #1389. Among other things I want to try adding a cache in a future PR, which should fix the bulk of the issues, but I still need to figure out a framework for more thorough profiling to make sure I'm optimizing the right things.

Altogether there shouldn't be more than a couple dozen megabytes of integrations content at any point, so I don't see any issues with adding an in-memory cache.